[Bug]: ClassCastException when using custom DynamicDestination in BigQueryIO.Write

### What happened?

**Beam: 2.40**

While using custom `DynamicDestination` in `BigQueryIO.Write` got the following exception:
```
java.lang.ClassCastException: org.apache.beam.sdk.io.gcp.bigquery.TableDestination cannot be cast to java.lang.String com.king.da.destinations.KingAppDestinations.getTable(KingAppDestinations.java:17)
 org.apache.beam.sdk.io.gcp.bigquery.UpdateSchemaDestination.processElement(UpdateSchemaDestination.java:131)
   org.apache.beam.sdk.io.gcp.bigquery.UpdateSchemaDestination$DoFnInvoker.invokeProcessElement(Unknown Source)
```
Find below test case to reproduce:

```java
import java.io.IOException;
import java.security.SecureRandom;
import java.util.Arrays;

import org.apache.beam.sdk.Pipeline;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.CreateDisposition;
import org.apache.beam.sdk.io.gcp.bigquery.BigQueryIO.Write.WriteDisposition;
import org.apache.beam.sdk.io.gcp.bigquery.DynamicDestinations;
import org.apache.beam.sdk.io.gcp.bigquery.TableDestination;
import org.apache.beam.sdk.io.gcp.bigquery.TestBigQueryOptions;
import org.apache.beam.sdk.io.gcp.testing.BigqueryClient;
import org.apache.beam.sdk.testing.TestPipeline;
import org.apache.beam.sdk.transforms.Create;
import org.apache.beam.sdk.values.KV;
import org.apache.beam.sdk.values.PCollection;
import org.apache.beam.sdk.values.ValueInSingleWindow;
import org.checkerframework.checker.nullness.qual.Nullable;
import org.junit.AfterClass;
import org.junit.BeforeClass;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

import com.google.api.services.bigquery.model.TableFieldSchema;
import com.google.api.services.bigquery.model.TableRow;
import com.google.api.services.bigquery.model.TableSchema;

@RunWith(JUnit4.class)
public class UpdateSchemaDestinationTest {

    static final BigqueryClient BQ_CLIENT = new BigqueryClient(UpdateSchemaDestinationTest.class.getName());

    static final String DATASET_ID =
            "schema_update_options_class_cast_excepption"
                + System.currentTimeMillis()
                + "_"
                + new SecureRandom().nextInt(32);

    static TestBigQueryOptions options = TestPipeline.testingPipelineOptions().as(TestBigQueryOptions.class);

    static Pipeline pipeline;

    @BeforeClass
    public static void setUpAll() throws IOException, InterruptedException {
        options.setTempLocation(options.getTempRoot() + "/bq_it_temp");

        pipeline = Pipeline.create(options);

        BQ_CLIENT.createNewDataset(options.getProject(), DATASET_ID);
    }

    @AfterClass
    public static void tearDownAll() {
        BQ_CLIENT.deleteDataset(options.getProject(), DATASET_ID);
    }

    @Test
    public void classCastExceptionRegression() {
        DynamicDestinations<KV<String, String>, String> destinations = new SomeDynamicDestinations();

        PCollection<KV<String, String>> rows = pipeline
            .apply(Create.of(
                KV.of("table","foo"),
                KV.of("table","bar")
            ));

        rows.apply(BigQueryIO.<KV<String, String>>write()
            .withFormatFunction(kv -> new TableRow().set("name", kv.getValue()))
            .withCreateDisposition(CreateDisposition.CREATE_IF_NEEDED)
            .withWriteDisposition(WriteDisposition.WRITE_APPEND)
            .withMaxBytesPerPartition(1)
            .to(destinations));

        pipeline.run().waitUntilFinish();
    }

    private static final class SomeDynamicDestinations extends DynamicDestinations<KV<String, String>, String> {

        private static final long serialVersionUID = 1L;

        @Override
        public String getDestination(@Nullable ValueInSingleWindow<KV<String, String>> element) {
            return element.getValue().getKey();
        }

        @Override
        public TableDestination getTable(String destination) {
            return new TableDestination(DATASET_ID + "." + destination, "a table");
        }

        @Override
        public @Nullable TableSchema getSchema(String destination) {
            return new TableSchema().setFields(Arrays.asList(new TableFieldSchema().setName("name").setType("STRING")));
        }
    }
}
```


### Issue Priority

Priority: 1

### Issue Component

Component: io-java-gcp

> ```java
> UpdateSchemaDestinationTest
> ```



> I am not able to replicate this exception with the code you provided.
> 
> I suspect the issue is that your dynamicDestinations expects a String argument to the getTable method, but a TableDestination is being passed instead, causing the casting exception


> I am not able to replicate this exception with the code you provided.

Are you getting any error or just green test? I am able to replicate the issue using the code I posted: 

```org.apache.beam.sdk.Pipeline$PipelineExecutionException: java.lang.ClassCastException: org.apache.beam.sdk.io.gcp.bigquery.TableDestination cannot be cast to java.lang.String
	at org.apache.beam.runners.direct.DirectRunner$DirectPipelineResult.waitUntilFinish(DirectRunner.java:373)
	at org.apache.beam.runners.direct.DirectRunner$DirectPipelineResult.waitUntilFinish(DirectRunner.java:341)
	at org.apache.beam.runners.direct.DirectRunner.run(DirectRunner.java:218)
	at org.apache.beam.runners.direct.DirectRunner.run(DirectRunner.java:67)
	at org.apache.beam.sdk.Pipeline.run(Pipeline.java:323)
	at org.apache.beam.sdk.Pipeline.run(Pipeline.java:309)
	at com.king.da.UpdateSchemaDestinationTest.classCastExceptionRegression(UpdateSchemaDestinationTest.java:78)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.junit.vintage.engine.execution.RunnerExecutor.execute(RunnerExecutor.java:42)
	at org.junit.vintage.engine.VintageTestEngine.executeAllChildren(VintageTestEngine.java:80)
	at org.junit.vintage.engine.VintageTestEngine.execute(VintageTestEngine.java:72)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:95)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:91)
	at org.junit.platform.launcher.core.SessionPerRequestLauncher.execute(SessionPerRequestLauncher.java:60)
	at org.eclipse.jdt.internal.junit5.runner.JUnit5TestReference.run(JUnit5TestReference.java:98)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:756)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
Caused by: java.lang.ClassCastException: org.apache.beam.sdk.io.gcp.bigquery.TableDestination cannot be cast to java.lang.String
	at com.king.da.UpdateSchemaDestinationTest$SomeDynamicDestinations.getTable(UpdateSchemaDestinationTest.java:1)
	at org.apache.beam.sdk.io.gcp.bigquery.UpdateSchemaDestination.processElement(UpdateSchemaDestination.java:131)
```

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
@@ -23,7 +23,6 @@ import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Prec
 
 import com.google.api.services.bigquery.model.TableReference;
 import com.google.api.services.bigquery.model.TableRow;
-import com.sun.xml.internal.bind.v2.TODO;
 import java.util.Collections;
 import java.util.List;
 import java.util.Set;
@@ -797,27 +796,31 @@ class BatchLoads<DestinationT, ElementT>
             .setCoder(KvCoder.of(destinationCoder, WriteTables.ResultCoder.INSTANCE));
 
     BigQueryOptions options = input.getPipeline().getOptions().as(BigQueryOptions.class);
-    String defaultProjectId = options.getBigQueryProject() == null
-        ? options.getProject()
-        : options.getBigQueryProject();
+    String defaultProjectId =
+        options.getBigQueryProject() == null ? options.getProject() : options.getBigQueryProject();
 
     return successfulWrites
         .apply(Keys.create())
-        .apply("Convert to TableDestinations", ParDo.of(new DoFn<DestinationT, TableDestination>() {
-          @ProcessElement
-          public void processElement(ProcessContext c) {
-            dynamicDestinations.setSideInputAccessorFromProcessContext(c);
-            TableDestination tableDestination = dynamicDestinations.getTable(c.element());
-            TableReference tableReference = tableDestination.getTableReference();
+        .apply(
+            "Convert to TableDestinations",
+            ParDo.of(
+                    new DoFn<DestinationT, TableDestination>() {
+                      @ProcessElement
+                      public void processElement(ProcessContext c) {
+                        dynamicDestinations.setSideInputAccessorFromProcessContext(c);
+                        TableDestination tableDestination =
+                            dynamicDestinations.getTable(c.element());
+                        TableReference tableReference = tableDestination.getTableReference();
 
-            // get project ID from options if it's not included in the table reference
-            if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
-              tableReference.setProjectId(defaultProjectId);
-              tableDestination = tableDestination.withTableReference(tableReference);
-            }
-            c.output(tableDestination);
-          }
-        }).withSideInputs(sideInputs))
+                        // get project ID from options if it's not included in the table reference
+                        if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
+                          tableReference.setProjectId(defaultProjectId);
+                          tableDestination = tableDestination.withTableReference(tableReference);
+                        }
+                        c.output(tableDestination);
+                      }
+                    })
+                .withSideInputs(sideInputs))
         .setCoder(tableDestinationCoder);
   }
 

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
@@ -21,6 +21,7 @@ import static org.apache.beam.sdk.io.gcp.bigquery.BigQueryHelpers.resolveTempLoc
 import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkArgument;
 import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Preconditions.checkState;
 
+import com.google.api.services.bigquery.model.TableReference;
 import com.google.api.services.bigquery.model.TableRow;
 import java.util.Collections;
 import java.util.List;
@@ -794,15 +795,26 @@ class BatchLoads<DestinationT, ElementT>
                     null))
             .setCoder(KvCoder.of(destinationCoder, WriteTables.ResultCoder.INSTANCE));
 
+    BigQueryOptions options = input.getPipeline().getOptions().as(BigQueryOptions.class);
+    String defaultProjectId = options.getBigQueryProject() == null
+        ? options.getProject()
+        : options.getBigQueryProject();
+
     return successfulWrites
         .apply(Keys.create())
-        .apply(
-            ParDo.of(
-                new DoFn<DestinationT, TableDestination>() {
-                  @ProcessElement
-                  public void processElement(
-                      @Element DestinationT dest, OutputReceiver<TableDestination> o) {
-                    o.output(dynamicDestinations.getTable(dest));
+        .apply("Convert to TableDestinations", MapElements.via(
+                new SimpleFunction<DestinationT, TableDestination>() {
+                  @Override
+                  public TableDestination apply(DestinationT dest) {
+                    TableDestination tableDestination = dynamicDestinations.getTable(dest);
+                    TableReference tableReference = tableDestination.getTableReference();
+
+                    // get project ID from options if it's not included in the table reference
+                    if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
+                      tableReference.setProjectId(defaultProjectId);
+                      tableDestination = tableDestination.withTableReference(tableReference);
+                    }
+                    return tableDestination;
                   }
                 }))
         .setCoder(tableDestinationCoder);

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
@@ -354,7 +354,7 @@ class BatchLoads<DestinationT, ElementT>
                             rowWriterFactory))
                     .withSideInputs(tempFilePrefixView)
                     .withOutputTags(multiPartitionsTag, TupleTagList.of(singlePartitionTag)));
-    PCollection<KV<TableDestination, WriteTables.Result>> tempTables =
+    PCollection<KV<DestinationT, WriteTables.Result>> tempTables =
         writeTempTables(partitions.get(multiPartitionsTag), tempLoadJobIdPrefixView);
 
     List<PCollectionView<?>> sideInputsForUpdateSchema =
@@ -366,7 +366,7 @@ class BatchLoads<DestinationT, ElementT>
             // Now that the load job has happened, we want the rename to happen immediately.
             .apply(
                 "Window Into Global Windows",
-                Window.<KV<TableDestination, WriteTables.Result>>into(new GlobalWindows())
+                Window.<KV<DestinationT, WriteTables.Result>>into(new GlobalWindows())
                     .triggering(Repeatedly.forever(AfterPane.elementCountAtLeast(1))))
             .apply("Add Void Key", WithKeys.of((Void) null))
             .setCoder(KvCoder.of(VoidCoder.of(), tempTables.getCoder()))
@@ -374,7 +374,7 @@ class BatchLoads<DestinationT, ElementT>
             .apply("Extract Values", Values.create())
             .apply(
                 ParDo.of(
-                        new UpdateSchemaDestination(
+                        new UpdateSchemaDestination<DestinationT>(
                             bigQueryServices,
                             tempLoadJobIdPrefixView,
                             loadJobProjectId,
@@ -473,7 +473,7 @@ class BatchLoads<DestinationT, ElementT>
             .apply("ReifyRenameInput", new ReifyAsIterable<>())
             .apply(
                 ParDo.of(
-                        new UpdateSchemaDestination(
+                        new UpdateSchemaDestination<DestinationT>(
                             bigQueryServices,
                             tempLoadJobIdPrefixView,
                             loadJobProjectId,
@@ -708,7 +708,7 @@ class BatchLoads<DestinationT, ElementT>
   }
 
   // Take in a list of files and write them to temporary tables.
-  private PCollection<KV<TableDestination, WriteTables.Result>> writeTempTables(
+  private PCollection<KV<DestinationT, WriteTables.Result>> writeTempTables(
       PCollection<KV<ShardedKey<DestinationT>, WritePartition.Result>> input,
       PCollectionView<String> jobIdTokenView) {
     List<PCollectionView<?>> sideInputs = Lists.newArrayList(jobIdTokenView);
@@ -718,9 +718,6 @@ class BatchLoads<DestinationT, ElementT>
         KvCoder.of(
             ShardedKeyCoder.of(NullableCoder.of(destinationCoder)),
             WritePartition.ResultCoder.INSTANCE);
-
-    Coder<TableDestination> tableDestinationCoder =
-        clusteringEnabled ? TableDestinationCoderV3.of() : TableDestinationCoderV2.of();
 
     // If WriteBundlesToFiles produced more than DEFAULT_MAX_FILES_PER_PARTITION files or
     // DEFAULT_MAX_BYTES_PER_PARTITION bytes, then
@@ -752,7 +749,7 @@ class BatchLoads<DestinationT, ElementT>
                 // https://github.com/apache/beam/issues/21105 for additional details.
                 schemaUpdateOptions,
                 tempDataset))
-        .setCoder(KvCoder.of(tableDestinationCoder, WriteTables.ResultCoder.INSTANCE));
+        .setCoder(KvCoder.of(destinationCoder, WriteTables.ResultCoder.INSTANCE));
   }
 
   // In the case where the files fit into a single load job, there's no need to write temporary
@@ -771,7 +768,7 @@ class BatchLoads<DestinationT, ElementT>
             ShardedKeyCoder.of(NullableCoder.of(destinationCoder)),
             WritePartition.ResultCoder.INSTANCE);
     // Write single partition to final table
-    PCollection<KV<TableDestination, WriteTables.Result>> successfulWrites =
+    PCollection<KV<DestinationT, WriteTables.Result>> successfulWrites =
         input
             .setCoder(partitionsCoder)
             // Reshuffle will distribute this among multiple workers, and also guard against
@@ -795,9 +792,20 @@ class BatchLoads<DestinationT, ElementT>
                     useAvroLogicalTypes,
                     schemaUpdateOptions,
                     null))
-            .setCoder(KvCoder.of(tableDestinationCoder, WriteTables.ResultCoder.INSTANCE));
+            .setCoder(KvCoder.of(destinationCoder, WriteTables.ResultCoder.INSTANCE));
 
-    return successfulWrites.apply(Keys.create());
+    return successfulWrites
+        .apply(Keys.create())
+        .apply(
+            ParDo.of(
+                new DoFn<DestinationT, TableDestination>() {
+                  @ProcessElement
+                  public void processElement(
+                      @Element DestinationT dest, OutputReceiver<TableDestination> o) {
+                    o.output(dynamicDestinations.getTable(dest));
+                  }
+                }))
+        .setCoder(tableDestinationCoder);
   }
 
   private WriteResult writeResult(Pipeline p, PCollection<TableDestination> successfulWrites) {

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BatchLoads.java
@@ -23,6 +23,7 @@ import static org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Prec
 
 import com.google.api.services.bigquery.model.TableReference;
 import com.google.api.services.bigquery.model.TableRow;
+import com.sun.xml.internal.bind.v2.TODO;
 import java.util.Collections;
 import java.util.List;
 import java.util.Set;
@@ -802,21 +803,21 @@ class BatchLoads<DestinationT, ElementT>
 
     return successfulWrites
         .apply(Keys.create())
-        .apply("Convert to TableDestinations", MapElements.via(
-                new SimpleFunction<DestinationT, TableDestination>() {
-                  @Override
-                  public TableDestination apply(DestinationT dest) {
-                    TableDestination tableDestination = dynamicDestinations.getTable(dest);
-                    TableReference tableReference = tableDestination.getTableReference();
+        .apply("Convert to TableDestinations", ParDo.of(new DoFn<DestinationT, TableDestination>() {
+          @ProcessElement
+          public void processElement(ProcessContext c) {
+            dynamicDestinations.setSideInputAccessorFromProcessContext(c);
+            TableDestination tableDestination = dynamicDestinations.getTable(c.element());
+            TableReference tableReference = tableDestination.getTableReference();
 
-                    // get project ID from options if it's not included in the table reference
-                    if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
-                      tableReference.setProjectId(defaultProjectId);
-                      tableDestination = tableDestination.withTableReference(tableReference);
-                    }
-                    return tableDestination;
-                  }
-                }))
+            // get project ID from options if it's not included in the table reference
+            if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
+              tableReference.setProjectId(defaultProjectId);
+              tableDestination = tableDestination.withTableReference(tableReference);
+            }
+            c.output(tableDestination);
+          }
+        }).withSideInputs(sideInputs))
         .setCoder(tableDestinationCoder);
   }
 

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/UpdateSchemaDestination.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/UpdateSchemaDestination.java
@@ -18,6 +18,7 @@
 package org.apache.beam.sdk.io.gcp.bigquery;
 
 import com.google.api.client.http.ByteArrayContent;
+import com.google.api.client.util.Strings;
 import com.google.api.services.bigquery.model.Clustering;
 import com.google.api.services.bigquery.model.EncryptionConfiguration;
 import com.google.api.services.bigquery.model.JobConfigurationLoad;
@@ -115,13 +116,29 @@ public class UpdateSchemaDestination<DestinationT>
     pendingJobs.clear();
   }
 
+  TableDestination getTableWithDefaultProject(DestinationT destination, BigQueryOptions options) {
+    TableDestination tableDestination = dynamicDestinations.getTable(destination);
+    TableReference tableReference = tableDestination.getTableReference();
+
+    if (Strings.isNullOrEmpty(tableReference.getProjectId())) {
+      tableReference.setProjectId(
+          options.getBigQueryProject() == null
+              ? options.getProject()
+              : options.getBigQueryProject());
+      tableDestination = tableDestination.withTableReference(tableReference);
+    }
+
+    return tableDestination;
+  }
+
   @ProcessElement
   public void processElement(
       @Element Iterable<KV<DestinationT, WriteTables.Result>> element,
       ProcessContext context,
       BoundedWindow window)
       throws IOException {
-    Object destination = null;
+    DestinationT destination = null;
+    BigQueryOptions options = context.getPipelineOptions().as(BigQueryOptions.class);
     for (KV<DestinationT, WriteTables.Result> entry : element) {
       destination = entry.getKey();
       if (destination != null) {
@@ -129,7 +146,7 @@ public class UpdateSchemaDestination<DestinationT>
       }
     }
     if (destination != null) {
-      TableDestination tableDestination = dynamicDestinations.getTable(destination);
+      TableDestination tableDestination = getTableWithDefaultProject(destination, options);
       TableSchema schema = dynamicDestinations.getSchema(destination);
       TableReference tableReference = tableDestination.getTableReference();
       String jobIdPrefix =
@@ -157,7 +174,7 @@ public class UpdateSchemaDestination<DestinationT>
     }
     List<KV<TableDestination, WriteTables.Result>> tableDestinations = new ArrayList<>();
     for (KV<DestinationT, WriteTables.Result> entry : element) {
-      tableDestinations.add(KV.of(dynamicDestinations.getTable(entry.getKey()), entry.getValue()));
+      tableDestinations.add(KV.of(getTableWithDefaultProject(destination, options), entry.getValue()));
     }
     context.output(tableDestinations);
   }

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/UpdateSchemaDestination.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/UpdateSchemaDestination.java
@@ -18,7 +18,6 @@
 package org.apache.beam.sdk.io.gcp.bigquery;
 
 import com.google.api.client.http.ByteArrayContent;
-import com.google.api.client.util.Strings;
 import com.google.api.services.bigquery.model.Clustering;
 import com.google.api.services.bigquery.model.EncryptionConfiguration;
 import com.google.api.services.bigquery.model.JobConfigurationLoad;
@@ -40,6 +39,7 @@ import org.apache.beam.sdk.transforms.DoFn;
 import org.apache.beam.sdk.transforms.windowing.BoundedWindow;
 import org.apache.beam.sdk.values.KV;
 import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.base.Strings;
 import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.Lists;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
@@ -174,7 +174,8 @@ public class UpdateSchemaDestination<DestinationT>
     }
     List<KV<TableDestination, WriteTables.Result>> tableDestinations = new ArrayList<>();
     for (KV<DestinationT, WriteTables.Result> entry : element) {
-      tableDestinations.add(KV.of(getTableWithDefaultProject(destination, options), entry.getValue()));
+      tableDestinations.add(
+          KV.of(getTableWithDefaultProject(destination, options), entry.getValue()));
     }
     context.output(tableDestinations);
   }

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/UpdateSchemaDestination.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/UpdateSchemaDestination.java
@@ -27,6 +27,7 @@ import com.google.api.services.bigquery.model.TableReference;
 import com.google.api.services.bigquery.model.TableSchema;
 import com.google.api.services.bigquery.model.TimePartitioning;
 import java.io.IOException;
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Set;
 import java.util.stream.Collectors;
@@ -43,9 +44,9 @@ import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 @SuppressWarnings({"nullness", "rawtypes"})
-public class UpdateSchemaDestination
+public class UpdateSchemaDestination<DestinationT>
     extends DoFn<
-        Iterable<KV<TableDestination, WriteTables.Result>>,
+        Iterable<KV<DestinationT, WriteTables.Result>>,
         Iterable<KV<TableDestination, WriteTables.Result>>> {
 
   private static final Logger LOG = LoggerFactory.getLogger(UpdateSchemaDestination.class);
@@ -116,12 +117,12 @@ public class UpdateSchemaDestination
 
   @ProcessElement
   public void processElement(
-      @Element Iterable<KV<TableDestination, WriteTables.Result>> element,
+      @Element Iterable<KV<DestinationT, WriteTables.Result>> element,
       ProcessContext context,
       BoundedWindow window)
       throws IOException {
     Object destination = null;
-    for (KV<TableDestination, WriteTables.Result> entry : element) {
+    for (KV<DestinationT, WriteTables.Result> entry : element) {
       destination = entry.getKey();
       if (destination != null) {
         break;
@@ -153,8 +154,12 @@ public class UpdateSchemaDestination
       if (updateSchemaDestinationJob != null) {
         pendingJobs.add(new PendingJobData(updateSchemaDestinationJob, tableDestination, window));
       }
-      context.output(element);
     }
+    List<KV<TableDestination, WriteTables.Result>> tableDestinations = new ArrayList<>();
+    for (KV<DestinationT, WriteTables.Result> entry : element) {
+      tableDestinations.add(KV.of(dynamicDestinations.getTable(entry.getKey()), entry.getValue()));
+    }
+    context.output(tableDestinations);
   }
 
   @Teardown

--- a/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java
+++ b/sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/WriteTables.java
@@ -97,7 +97,7 @@ import org.slf4j.LoggerFactory;
 class WriteTables<DestinationT extends @NonNull Object>
     extends PTransform<
         PCollection<KV<ShardedKey<DestinationT>, WritePartition.Result>>,
-        PCollection<KV<TableDestination, WriteTables.Result>>> {
+        PCollection<KV<DestinationT, WriteTables.Result>>> {
   @AutoValue
   abstract static class Result {
     abstract String getTableName();
@@ -135,7 +135,7 @@ class WriteTables<DestinationT extends @NonNull Object>
   private final Set<SchemaUpdateOption> schemaUpdateOptions;
   private final DynamicDestinations<?, DestinationT> dynamicDestinations;
   private final List<PCollectionView<?>> sideInputs;
-  private final TupleTag<KV<TableDestination, WriteTables.Result>> mainOutputTag;
+  private final TupleTag<KV<DestinationT, WriteTables.Result>> mainOutputTag;
   private final TupleTag<String> temporaryFilesTag;
   private final @Nullable ValueProvider<String> loadJobProjectId;
   private final int maxRetryJobs;
@@ -148,8 +148,7 @@ class WriteTables<DestinationT extends @NonNull Object>
   private final @Nullable String tempDataset;
 
   private class WriteTablesDoFn
-      extends DoFn<
-          KV<ShardedKey<DestinationT>, WritePartition.Result>, KV<TableDestination, Result>> {
+      extends DoFn<KV<ShardedKey<DestinationT>, WritePartition.Result>, KV<DestinationT, Result>> {
 
     private Map<DestinationT, String> jsonSchemas = Maps.newHashMap();
 
@@ -160,6 +159,7 @@ class WriteTables<DestinationT extends @NonNull Object>
       final List<String> partitionFiles;
       final TableDestination tableDestination;
       final TableReference tableReference;
+      final DestinationT destinationT;
       final boolean isFirstPane;
 
       public PendingJobData(
@@ -168,12 +168,14 @@ class WriteTables<DestinationT extends @NonNull Object>
           List<String> partitionFiles,
           TableDestination tableDestination,
           TableReference tableReference,
+          DestinationT destinationT,
           boolean isFirstPane) {
         this.window = window;
         this.retryJob = retryJob;
         this.partitionFiles = partitionFiles;
         this.tableDestination = tableDestination;
         this.tableReference = tableReference;
+        this.destinationT = destinationT;
         this.isFirstPane = isFirstPane;
       }
     }
@@ -292,6 +294,7 @@ class WriteTables<DestinationT extends @NonNull Object>
               partitionFiles,
               tableDestination,
               tableReference,
+              destination,
               element.getValue().isFirstPane()));
     }
 
@@ -359,7 +362,7 @@ class WriteTables<DestinationT extends @NonNull Object>
                             pendingJob.isFirstPane);
                     c.output(
                         mainOutputTag,
-                        KV.of(pendingJob.tableDestination, result),
+                        KV.of(pendingJob.destinationT, result),
                         pendingJob.window.maxTimestamp(),
                         pendingJob.window);
                     for (String file : pendingJob.partitionFiles) {
@@ -423,7 +426,7 @@ class WriteTables<DestinationT extends @NonNull Object>
   }
 
   @Override
-  public PCollection<KV<TableDestination, Result>> expand(
+  public PCollection<KV<DestinationT, Result>> expand(
       PCollection<KV<ShardedKey<DestinationT>, WritePartition.Result>> input) {
     PCollectionTuple writeTablesOutputs =
         input.apply(

--- a/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java
+++ b/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java
@@ -104,8 +104,10 @@ import org.apache.beam.sdk.testing.TestPipeline;
 import org.apache.beam.sdk.testing.TestStream;
 import org.apache.beam.sdk.transforms.Create;
 import org.apache.beam.sdk.transforms.Distinct;
+import org.apache.beam.sdk.transforms.DoFn;
 import org.apache.beam.sdk.transforms.DoFnTester;
 import org.apache.beam.sdk.transforms.MapElements;
+import org.apache.beam.sdk.transforms.ParDo;
 import org.apache.beam.sdk.transforms.SerializableFunction;
 import org.apache.beam.sdk.transforms.SerializableFunctions;
 import org.apache.beam.sdk.transforms.SimpleFunction;
@@ -2230,25 +2232,31 @@ public class BigQueryIOWriteTest implements Serializable {
             Collections.emptySet(),
             null);
 
-    PCollection<KV<String, WriteTables.Result>> writeTablesOutput =
+    PCollection<KV<TableDestination, WriteTables.Result>> writeTablesOutput =
         writeTablesInput
             .apply(writeTables)
-            .setCoder(KvCoder.of(StringUtf8Coder.of(), WriteTables.ResultCoder.INSTANCE));
+            .setCoder(KvCoder.of(StringUtf8Coder.of(), WriteTables.ResultCoder.INSTANCE))
+            .apply(ParDo.of(new DoFn<KV<String, WriteTables.Result>, KV<TableDestination, WriteTables.Result>>() {
+              @ProcessElement
+              public void processElement(@Element KV<String, WriteTables.Result> e, OutputReceiver<KV<TableDestination, WriteTables.Result>> o) {
+                o.output(KV.of(dynamicDestinations.getTable(e.getKey()), e.getValue()));
+              }
+            }));
+
+
 
     PAssert.thatMultimap(writeTablesOutput)
         .satisfies(
             input -> {
-              assertEquals(input.keySet(), expectedTempTables.keySet());
-              for (Map.Entry<String, Iterable<WriteTables.Result>> entry : input.entrySet()) {
+              assertEquals(expectedTempTables.keySet(), input.keySet());
+              for (Map.Entry<TableDestination, Iterable<WriteTables.Result>> entry : input.entrySet()) {
                 Iterable<String> tableNames =
                     StreamSupport.stream(entry.getValue().spliterator(), false)
                         .map(Result::getTableName)
                         .collect(Collectors.toList());
                 @SuppressWarnings("unchecked")
                 String[] expectedValues =
-                    Iterables.toArray(
-                        expectedTempTables.get(dynamicDestinations.getTable(entry.getKey())),
-                        String.class);
+                    Iterables.toArray(expectedTempTables.get(entry.getKey()), String.class);
                 assertThat(tableNames, containsInAnyOrder(expectedValues));
               }
               return null;

--- a/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java
+++ b/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java
@@ -2236,20 +2236,25 @@ public class BigQueryIOWriteTest implements Serializable {
         writeTablesInput
             .apply(writeTables)
             .setCoder(KvCoder.of(StringUtf8Coder.of(), WriteTables.ResultCoder.INSTANCE))
-            .apply(ParDo.of(new DoFn<KV<String, WriteTables.Result>, KV<TableDestination, WriteTables.Result>>() {
-              @ProcessElement
-              public void processElement(@Element KV<String, WriteTables.Result> e, OutputReceiver<KV<TableDestination, WriteTables.Result>> o) {
-                o.output(KV.of(dynamicDestinations.getTable(e.getKey()), e.getValue()));
-              }
-            }));
-
-
+            .apply(
+                ParDo.of(
+                    new DoFn<
+                        KV<String, WriteTables.Result>,
+                        KV<TableDestination, WriteTables.Result>>() {
+                      @ProcessElement
+                      public void processElement(
+                          @Element KV<String, WriteTables.Result> e,
+                          OutputReceiver<KV<TableDestination, WriteTables.Result>> o) {
+                        o.output(KV.of(dynamicDestinations.getTable(e.getKey()), e.getValue()));
+                      }
+                    }));
 
     PAssert.thatMultimap(writeTablesOutput)
         .satisfies(
             input -> {
               assertEquals(expectedTempTables.keySet(), input.keySet());
-              for (Map.Entry<TableDestination, Iterable<WriteTables.Result>> entry : input.entrySet()) {
+              for (Map.Entry<TableDestination, Iterable<WriteTables.Result>> entry :
+                  input.entrySet()) {
                 Iterable<String> tableNames =
                     StreamSupport.stream(entry.getValue().spliterator(), false)
                         .map(Result::getTableName)

--- a/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java
+++ b/sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOWriteTest.java
@@ -2209,6 +2209,8 @@ public class BigQueryIOWriteTest implements Serializable {
         p.apply("CreateJobId", Create.of("jobId")).apply(View.asSingleton());
     List<PCollectionView<?>> sideInputs = ImmutableList.of(jobIdTokenView);
 
+    DynamicDestinations<String, String> dynamicDestinations = new IdentityDynamicTables();
+
     fakeJobService.setNumFailuresExpected(3);
     WriteTables<String> writeTables =
         new WriteTables<>(
@@ -2218,7 +2220,7 @@ public class BigQueryIOWriteTest implements Serializable {
             BigQueryIO.Write.WriteDisposition.WRITE_EMPTY,
             BigQueryIO.Write.CreateDisposition.CREATE_IF_NEEDED,
             sideInputs,
-            new IdentityDynamicTables(),
+            dynamicDestinations,
             null,
             4,
             false,
@@ -2228,24 +2230,25 @@ public class BigQueryIOWriteTest implements Serializable {
             Collections.emptySet(),
             null);
 
-    PCollection<KV<TableDestination, WriteTables.Result>> writeTablesOutput =
+    PCollection<KV<String, WriteTables.Result>> writeTablesOutput =
         writeTablesInput
             .apply(writeTables)
-            .setCoder(KvCoder.of(TableDestinationCoderV3.of(), WriteTables.ResultCoder.INSTANCE));
+            .setCoder(KvCoder.of(StringUtf8Coder.of(), WriteTables.ResultCoder.INSTANCE));
 
     PAssert.thatMultimap(writeTablesOutput)
         .satisfies(
             input -> {
               assertEquals(input.keySet(), expectedTempTables.keySet());
-              for (Map.Entry<TableDestination, Iterable<WriteTables.Result>> entry :
-                  input.entrySet()) {
+              for (Map.Entry<String, Iterable<WriteTables.Result>> entry : input.entrySet()) {
                 Iterable<String> tableNames =
                     StreamSupport.stream(entry.getValue().spliterator(), false)
                         .map(Result::getTableName)
                         .collect(Collectors.toList());
                 @SuppressWarnings("unchecked")
                 String[] expectedValues =
-                    Iterables.toArray(expectedTempTables.get(entry.getKey()), String.class);
+                    Iterables.toArray(
+                        expectedTempTables.get(dynamicDestinations.getTable(entry.getKey())),
+                        String.class);
                 assertThat(tableNames, containsInAnyOrder(expectedValues));
               }
               return null;

